javarosa-dag.rst

```diff
@@ -0,0 +1,304 @@
+Those expressions can be put in the following attributes of the `bind <https://opendatakit.github.io/xforms-spec/#bindings>`_ section 
+in a form:
+
+* ``calculate``
+* ``relevant``
+* ``required``
+* ``readonly``
+
+The ``constraint`` attribute also takes an XPath expression but it is not managed by DAG.
+It is represented by ``Constraint`` class and evaluated before an answer is committed into the data model
+in the ``FormEntryController.answerQuestion(FormIndex index, IAnswerData data, boolean midSurvey)`` method.
+
+Classes
+-------
+
+- ``Triggerable`` --- An abstract class represents such expression, its context and targets which should be updated by this expression. It has two implementations:
+
+    * ``Condition`` --- Represents an expression for ``relevant``, ``required`` and ``readonly`` attributes which always evaluates to ``true`` or ``false``.
+    * ``Recalculate``--- Represents an expression for the ``calculate`` attribute and can evaluate to any supported data type.
+- ``QuickTriggerable`` --- A ``Triggerable`` class wrapper which provides a shallow comparison of ``Triggerble`` implementations. Once the DAG structure is built, no new ``Triggerable`` instances are created and therefore there is no need for the deep comparison. For the sake of readability, this document will always reference to ``Triggerable`` class even if ``QuickTriggerable`` is actually used in the code. 
+- ``IDag`` --- An abstract class represents an interface for the DAG management and triggerable evaluation logic.
+- ``LatestDagBase`` --- An abstract implementation of ``IDag`` class which serves as a base for all actual implementation of ``IDag``
+- ``Safe2014DagImpl`` --- The only supported ``IDag`` implementation.
+- ``FormDef`` -- A form representation along with the data model. This class communicates with ``IDag``.
+
+High level description
+-----------------------
+
+Each triggerable expression is evaluated at form load and save. Additionally, if the XPath expression references other node(s),
+it is evaluated if any of the referenced node is updated. Moreover, it may happen that a referenced node also has a triggerable
+expression which references other nodes etc. 
+So triggerables may depend on each other. 
+
+Lets check an example:
+Assume there are nodes ``nodeA``, ``nodeB`` and ``nodeC`` and:
+
+* ``nodeA`` has a ``calculate`` expression ``tA`` which references the ``nodeB`` value
+* ``nodeB`` has a ``calculate`` expression ``tB`` which references the ``nodeC`` value
+* ``nodeC`` has no triggerable expressions
+
+So the dependency is ``tC`` > ``tB`` > ``tA``.
+When a form is loaded or saved, those triggerables should be evaluated in the following order: ``tC``, ``tB``, ``tA``.
+When a value of ``nodeC`` is changed then ``tB`` and ``tA`` must be evaluated. In this order.
+
+Lets change the above so the ``nodeC`` node has a ``relevant`` expression which returns ``true`` if ``nodeA`` has certain value.
```

Comment:
Sorry for all the copy editing here. Perhaps I’ll also have something substantive to add.

>has certain

has a certain